core/src/serve_repair.rs

```diff
@@ -1027,11 +1027,17 @@ impl ServeRepair {
                 continue;
             }
             if let Ok(RepairResponse::Ping(ping)) = packet.deserialize_slice(..) {
-                packet.meta.set_discard(true);
                 if !ping.verify() {
+                    // Do _not_ set `discard` to allow shred processing to attempt to
+                    // handle the packet.
+                    // Ping error count may include false posities for shreds of size
+                    // `REPAIR_RESPONSE_SERIALIZED_PING_BYTES` whose first 4 bytes
+                    // match `RepairResponse` discriminator (these 4 bytes overlap
+                    // with the shred signature field).
                     stats.ping_err_verify_count += 1;
                     continue;
                 }
+                packet.meta.set_discard(true);
```

Comment:
isn't it still _theoretically_ possible to have a valid shred which is erroneously ingested as a ping packet here?

Reply:
Theoretically, if the signature verifies at the appropriate offsets.

The shred would have to be formatted in such a way that the size matches the ping response size, the first four bytes of the shred signature map to a `RepairResponse` discriminator, and the buffer which spans part of the shred signature, shred variant and shred slot is a valid signature for the following `Hash` size buffer offset into the shred header and data.